### PR TITLE
OCM-7254 | feat: Move edit machinepool non-cmd funcs to pkg and test

### PR DESCRIPTION
**Pull request by hunterkepley**
[OCM-7254](https://issues.redhat.com//browse/OCM-7254)

* Move non-cmd funcs from cmd/edit/machinepool to pkg/machinepool
* Test
* Refactor to be used in pkg instead of cmd

### CODE REVIEW

**Review by willkutler — APPROVED** [review]

File: cmd/edit/machinepool/cmd_test.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package machinepool
+
+import (
+	. "github.com/onsi/ginkgo/v2"
+)
+
+// Will un-comment and become the cmd test once the work on refactoring this command is finalized
+var _ = Describe("Edit Machinepool", func() {
```
Comment (hunterkepley) [inline cmd/edit/machinepool/cmd_test.go:8]: Moved all tests in here to `pkg/machinepool/machinepool_test.go` -- except the one commented out here as it tests the actual command, which will be rewritten in the next MR immediately following this one.

File: pkg/machinepool/machinepool.go [hunk 2]
```diff
@@ -280,3 +291,691 @@ func getNodePoolsString(nodePools []*cmv1.NodePool) string {
 	}
 	return outputString
 }
+
```
Comment (hunterkepley) [inline pkg/machinepool/machinepool.go:294]: Here begins the moved functions from `cmd/edit/machinepool`

File: pkg/machinepool/machinepool.go [hunk 3]
```diff
@@ -280,3 +291,691 @@ func getNodePoolsString(nodePools []*cmv1.NodePool) string {
...
+	}
+
+	if existingMinReplica != minReplicas || existingMaxReplica != maxReplicas {
+		if min >= 1 && max >= 1 {
+			return cmv1.NewNodePoolAutoscaling().MinReplica(min).MaxReplica(max)
+		}
+	}
+
+	return nil
+}
```
Comment (hunterkepley) [inline pkg/machinepool/machinepool.go:981]: I made a couple of changes, which was just grabbing the args via cobra flags, as this was the first refactor which used the args directly in these functions. Otherwise, we have just been grabbing machinepool/nodepool ID + cluster in the args, which is included in the runtime + parameters for each function

If I include all the args for `edit machinepool`, the parameters for the functions would be quite long.

I debated making a struct to contain this information, but this file is already 1000 lines long, so i am at a crossroads:

1. no struct, keep as is (using cmd.flag("FLAG_NAME").value to get the arguments)
2. separate nodepool and machinepool funcs into their own files (preferred) and use a struct (potentially a 3rd file for data types)

File: pkg/machinepool/machinepool_test.go [hunk 4]
```diff
@@ -26,6 +28,112 @@ var _ = Describe("Machinepool and nodepool", func() {
 				Value(cmv1.UpgradePolicyStateValueScheduled)).
 				NextRun(date)
 		})
+		It("editAutoscaling should equal nil if nothing is changed", func() {
```
Comment (hunterkepley) [inline pkg/machinepool/machinepool_test.go:31]: Tests moved here, separated by machinepool & nodepool. Taken from the original `cmd/edit/machinepool` tests, and modified to fit the refactor

**Comment by robpblake** [comment]
Holding for cut of `1.2.41` release.

File: pkg/machinepool/machinepool.go (hunk 3 above)
Comment (ciaranRoche) [reply pkg/machinepool/machinepool.go:1035]: lets not make this PR any bigger. Keep as is for now. If you want to follow up with your suggestion in a different MR that would be good

File: Makefile [hunk 5]
```diff
@@ -38,7 +38,7 @@ test:
 
 .PHONY: coverage
 coverage:
-	go test -coverprofile=cover.out -covermode=atomic -p 4 $(shell go list ./... | grep -v /tests/)
+	go test -coverprofile=cover.out -coverpkg="./..." -covermode=atomic -p 4 $(shell go list ./... | grep -v /tests/)
```
Comment (robpblake) [inline Makefile:41]: @hunterkepley I would say lets revert this change as it doesn't seem to be doing what we expect.

**Review by ciaranRoche — COMMENTED** [review]
Overall looks good, lets not lets follow up with the changes you have suggested in another MR

File: cmd/edit/machinepool/cmd.go [hunk 6]
```diff
@@ -180,12 +181,8 @@ func run(cmd *cobra.Command, argv []string) {
 		}
 	}
 
-	var err error
-	if cluster.Hypershift().Enabled() {
-		err = editNodePool(cmd, machinePoolID, clusterKey, cluster, r)
-	} else {
-		err = editMachinePool(cmd, machinePoolID, clusterKey, cluster, r)
-	}
+	service := machinepool.NewMachinePoolService()
```
Comment (robpblake) [inline cmd/edit/machinepool/cmd.go:184]: @hunterkepley Not for this PR, but I've been thinking it would be good for us to have a pattern to be able to mock out the services that are created like this with an uber mock. It would make testing the `runner` much easier. My initial thought is to put the service onto the cmd options that get passed in so for testing we can explicitly set the mock, and for normal execution, we just set the correct implementation or have a pattern whereby if the service is nil, we create one.

File: cmd/create/machinepool/nodepool.go [hunk 7]
```diff
@@ -472,13 +472,11 @@ func addNodePool(cmd *cobra.Command, clusterKey string, cluster *cmv1.Cluster, r
...
-				Default:  maxSurge,
-				Required: false,
-				Validators: []interactive.Validator{
-					machinepools.ValidateUpgradeMaxSurgeUnavailable,
-				},
+				Question:   "Max surge",
+				Help:       cmd.Flags().Lookup("max-surge").Usage,
+				Default:    maxSurge,
+				Required:   false,
+				Validators: []interactive.Validator{},
```
Comment (marcolan018) [inline cmd/create/machinepool/nodepool.go:479]: just want to confirm: ValidateUpgradeMaxSurgeUnavailable is not required?

Comment (hunterkepley) [reply cmd/create/machinepool/nodepool.go:479]: I think i removed by mistake, re-adding, thanks marco

Comment (hunterkepley) [reply cmd/create/machinepool/nodepool.go]: Ah, i think this was from a rebase, as it's not related to this MR, and I must have kept the wrong version. Either way fixed now :)

File: cmd/describe/machinepool/cmd_test.go [hunk 8]
```diff
@@ -38,10 +38,6 @@ Tuning configs:
 Kubelet configs:                       
 Additional security group IDs:         
 Node drain grace period:               1 minute
-Management upgrade:
```
Comment (robpblake) [inline cmd/describe/machinepool/cmd_test.go:41]: @hunterkepley Is this delete expected?

File: pkg/machinepool/output.go [hunk 9]
```diff
@@ -82,7 +81,6 @@ func nodePoolOutput(clusterId string, nodePool *cmv1.NodePool) string {
 		ocmOutput.PrintNodePoolConfigs(nodePool.KubeletConfigs()),
 		ocmOutput.PrintNodePoolAdditionalSecurityGroups(nodePool.AWSNodePool()),
 		ocmOutput.PrintNodeDrainGracePeriod(nodePool.NodeDrainGracePeriod()),
-		ocmOutput.PrintNodePoolManagementUpgrade(nodePool.ManagementUpgrade()),
```
Comment (robpblake) [inline pkg/machinepool/output.go:85]: @hunterkepley Is it expected to be removing these from the output?

File: pkg/ocm/output/nodepools.go [hunk 10]
```diff
@@ -115,14 +115,3 @@ func PrintNodeDrainGracePeriod(period *cmv1.Value) string {
 
 	return ""
 }
-
-func PrintNodePoolManagementUpgrade(upgrade *cmv1.NodePoolManagementUpgrade) string {
```
Comment (robpblake) [inline pkg/ocm/output/nodepools.go:119]: @hunterkepley To confirm again: This is expected?

File: pkg/helper/machinepools/helpers_test.go [hunk 11]
```diff
@@ -330,57 +330,3 @@ var _ = Describe("Validate node drain grace period", func() {
 		),
 	)
 })
-
-var _ = Describe("Validate MaxSurge and MaxUnavailable", func() {
```
Comment (robpblake) [inline pkg/helper/machinepools/helpers_test.go:334]: @hunterkepley I don't see this test being replicated anywhere. It looks like it's just been deleted?

File: pkg/ocm/output/nodepools_test.go [hunk 12]
```diff
@@ -46,20 +44,3 @@ var _ = Describe("PrintNodePoolReplicasInline", func() {
 	})
 
 })
-
-var _ = Describe("Validate management upgrade print output", func() {
```
Comment (robpblake) [inline pkg/ocm/output/nodepools_test.go:50]: @hunterkepley I don't see this test being replicated. It looks like it's just being deleted?

File: pkg/machinepool/machinepool_test.go [hunk 13]
```diff
@@ -26,6 +28,112 @@ var _ = Describe("Machinepool and nodepool", func() {
...
+			})
+
+			It("Prompts when the user has deleted a kubelet-config", func() {
+
+				invoked := false
+
+				f := func(r *rosa.Runtime) bool {
+					invoked = true
+					return invoked
+				}
```
Comment (ziccardi) [inline pkg/machinepool/machinepool_test.go:81]: DRY
This is repeated for each test in this context.
That can be avoided adding this code:
```go
var mockPromptFuncInvoked bool

BeforeEach(func() {
	t = NewTestRuntime()
	mockPromptFuncInvoked = false
})

mockPromptfunc := func(r *rosa.Runtime) bool {
	mockPromptFuncInvoked = true
	return mockPromptFuncInvoked
}
```
Then you can pass `mockPromptFuncInvoked` everywhere `f` is passed and check the value f `mockPromptFuncInvoked` instead of `invoked`

File: pkg/machinepool/machinepool_test.go [hunk 14]
```diff
@@ -26,6 +28,107 @@ var _ = Describe("Machinepool and nodepool", func() {
...
+					n.KubeletConfigs("test")
+				})
+
+				update := MockNodePool(func(n *cmv1.NodePoolBuilder) {
+					n.KubeletConfigs("")
+				})
+
+				Expect(promptForNodePoolNodeRecreate(original, update, invoked, t.RosaRuntime)).To(BeTrue())
+				Expect(mockPromptFuncInvoked).To(BeTrue())
+			})
```
Comment (ziccardi) [inline pkg/machinepool/machinepool_test.go:92]: DRY
The
```go
				invoked := func(r *rosa.Runtime) bool {
					mockPromptFuncInvoked = true
					return mockPromptFuncInvoked
				}
```
does not need to be repeated in each test

```suggestion
		Context("Prompt For NodePoolNodeRecreate", func() {
			var mockPromptFuncInvoked bool
			var t *TestingRuntime
			BeforeEach(func() {
				t = NewTestRuntime()
				mockPromptFuncInvoked = false
			})

			mockPromptFunc := func(r *rosa.Runtime) bool {
				mockPromptFuncInvoked = true
				return mockPromptFuncInvoked
			}


			It("Prompts when the user has deleted a kubelet-config", func() {
				original := MockNodePool(func(n *cmv1.NodePoolBuilder) {
					n.KubeletConfigs("test")
				})

				update := MockNodePool(func(n *cmv1.NodePoolBuilder) {
					n.KubeletConfigs("")
				})

				Expect(promptForNodePoolNodeRecreate(original, update, mockPromptFunc, t.RosaRuntime)).To(BeTrue())
				Expect(mockPromptFuncInvoked).To(BeTrue())
			})
```
The same `mockPromptFunc` can be used in each test.

File: cmd/edit/machinepool/cmd_test.go [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+	/*Context("getNodePoolReplicas", func() {
```
Comment (ciaranRoche) [inline cmd/edit/machinepool/cmd_test.go:10]: Do we want to leave in this commented out test?

Comment (hunterkepley) [reply cmd/edit/machinepool/cmd_test.go:10]: I would like to- theyre broken with the refactor, would like to do full end-to-end tests with the new runner/testing structure after the next MR which will introduce this new runner

File: pkg/machinepool/machinepool_test.go [hunk 16]
```diff
@@ -92,8 +184,95 @@ var _ = Describe("Machinepool and nodepool", func() {
...
+			var npBuilder *cmv1.NodePoolBuilder
+			existingNodepool, err := cmv1.NewNodePool().
+				Autoscaling(cmv1.NewNodePoolAutoscaling().MaxReplica(4).MinReplica(1)).
+				Build()
+			Expect(err).To(BeNil())
+			It("Autoscaling set", func() {
+				npBuilder = cmv1.NewNodePool()
+				fillAutoScalingAndReplicas(npBuilder, true, existingNodepool, 1, 3, 2)
+				npPatch, err := npBuilder.Build()
+				Expect(err).To(BeNil())
```
Comment (ziccardi) [inline pkg/machinepool/machinepool_test.go:197]:
```suggestion
				Expect(err).ToNot(HaveOccurred())
```

File: pkg/machinepool/machinepool_test.go [hunk 17]
```diff
@@ -92,8 +184,95 @@ var _ = Describe("Machinepool and nodepool", func() {
...
+				Expect(npPatch.Autoscaling()).ToNot(BeNil())
+				// Default (zero) value
+				Expect(npPatch.Replicas()).To(Equal(0))
+			})
+			It("Replicas set", func() {
+				npBuilder = cmv1.NewNodePool()
+				fillAutoScalingAndReplicas(npBuilder, false, existingNodepool, 0, 0, 2)
+				npPatch, err := npBuilder.Build()
+				Expect(err).To(BeNil())
```
Comment (ziccardi) [inline pkg/machinepool/machinepool_test.go:206]:
```suggestion
				Expect(err).ToNot(HaveOccurred())
```

File: pkg/machinepool/machinepool_test.go [hunk 18]
```diff
@@ -92,8 +184,95 @@ var _ = Describe("Machinepool and nodepool", func() {
...
+				Expect(err).ToNot(HaveOccurred())
+				boolean := isMultiAZMachinePool(machinePool)
+				Expect(boolean).To(Equal(true))
+			})
+
+			It("isMultiAZMachinePool should return false", func() {
+				machinePool, err := cmv1.NewMachinePool().AvailabilityZones("test").Build()
+				Expect(err).ToNot(HaveOccurred())
+				boolean := isMultiAZMachinePool(machinePool)
+				Expect(boolean).To(Equal(false))
```
Comment (ziccardi) [inline pkg/machinepool/machinepool_test.go:273]:
```suggestion
				Expect(boolean).To(BeFalse())
```

File: pkg/machinepool/machinepool_test.go [hunk 19]
```diff
@@ -92,8 +184,95 @@ var _ = Describe("Machinepool and nodepool", func() {
...
+				Expect(builder).To(Equal(asBuilder))
+			})
+		})
+
+		Context("isMultiAZMachinePool", func() {
+			It("isMultiAZMachinePool should return true", func() {
+				machinePool, err := cmv1.NewMachinePool().Build()
+				Expect(err).ToNot(HaveOccurred())
+				boolean := isMultiAZMachinePool(machinePool)
+				Expect(boolean).To(Equal(true))
```
Comment (ziccardi) [inline pkg/machinepool/machinepool_test.go:266]:
```suggestion
				Expect(boolean).To(BeTrue())
```

File: pkg/machinepool/machinepool_test.go [hunk 20]
```diff
@@ -92,8 +184,95 @@ var _ = Describe("Machinepool and nodepool", func() {
 			fmt.Println(out)
 			Expect(fmt.Sprint(out)).To(Equal(fmt.Sprint(expectedOutput)))
 		})
+		Context("fillAutoScalingAndReplicas", func() {
+			var npBuilder *cmv1.NodePoolBuilder
+			existingNodepool, err := cmv1.NewNodePool().
+				Autoscaling(cmv1.NewNodePoolAutoscaling().MaxReplica(4).MinReplica(1)).
+				Build()
+			Expect(err).To(BeNil())
```
Comment (ziccardi) [inline pkg/machinepool/machinepool_test.go:192]:
```suggestion
			Expect(err).ToNot(HaveOccurred())
```

**Review by robpblake — APPROVED** [review]
/approve
